File: agent/security/content/entropy.py
"""
Shannon Entropy Calculator
High entropy (> 7.0) indicates encrypted, compressed, or packed content.
Used to detect obfuscated malware payloads.
"""
import math
from collections import Counter


def shannon_entropy(data: bytes) -> float:
    """
    Calculate Shannon entropy of binary data.
    Returns value between 0.0 (uniform) and 8.0 (maximum randomness).
    """
    if not data:
        return 0.0

    counts = Counter(data)
    total = len(data)
    entropy = 0.0
    for count in counts.values():
        p = count / total
        if p > 0:
            entropy -= p * math.log2(p)
    return entropy


def is_suspicious_entropy(data: bytes, threshold: float = 7.2) -> tuple[bool, float]:
    """
    Returns (is_suspicious, score).
    Threshold 7.2 catches most packed/encrypted payloads while avoiding
    false positives on legitimate compressed assets.
    """
    score = shannon_entropy(data)
    return score >= threshold, score


def entropy_verdict(data: bytes) -> dict:
    """Full entropy analysis result."""
    score = shannon_entropy(data)
    suspicious = score >= 7.2
    level = "CLEAN"
    if score >= 7.8:
        level = "HIGH_ENTROPY_LIKELY_ENCRYPTED"
    elif score >= 7.2:
        level = "HIGH_ENTROPY_SUSPICIOUS"
    elif score >= 6.0:
        level = "MEDIUM_ENTROPY"
    return {
        "score": round(score, 4),
        "suspicious": suspicious,
        "level": level,
    }

File: internal/security/content/entropy.go
// Package content provides content inspection: entropy, magic bytes, Base64 decoding.
package content

import (
	"math"
)

const (
	EntropyHighThreshold = 7.2 // above = suspicious (encrypted/packed)
	EntropyVeryHigh      = 7.8 // above = likely encrypted
)

// EntropyResult describes the entropy analysis of a byte slice.
type EntropyResult struct {
	Score      float64
	Suspicious bool
	Level      string
}

// ShannonEntropy computes the Shannon entropy of binary data (0.0–8.0).
// Values above 7.0 strongly suggest encryption, compression, or packing.
func ShannonEntropy(data []byte) float64 {
	if len(data) == 0 {
		return 0
	}
	var freq [256]int
	for _, b := range data {
		freq[b]++
	}
	n := float64(len(data))
	var h float64
	for _, count := range freq {
		if count == 0 {
			continue
		}
		p := float64(count) / n
		h -= p * math.Log2(p)
	}
	return h
}

// AnalyseEntropy runs entropy analysis and classifies the result.
func AnalyseEntropy(data []byte) EntropyResult {
	score := ShannonEntropy(data)
	suspicious := score >= EntropyHighThreshold
	level := "CLEAN"
	switch {
	case score >= EntropyVeryHigh:
		level = "HIGH_ENTROPY_LIKELY_ENCRYPTED"
	case score >= EntropyHighThreshold:
		level = "HIGH_ENTROPY_SUSPICIOUS"
	case score >= 6.0:
		level = "MEDIUM_ENTROPY"
	}
	return EntropyResult{Score: math.Round(score*10000) / 10000, Suspicious: suspicious, Level: level}
}
